plugin/adsk/plugin/geomNode.cpp

```diff
@@ -0,0 +1,236 @@
+        , mBuilder(&dataBlock, attribute, size)
+    {
+    }
+
+    template <typename T> void add(const T& item)
+    {
+        MDataHandle element = mBuilder.addLast();
+        element.set(item);
+    }
+
+    void finish() { mArrayDataHandle.set(mBuilder); }
+
+private:
+    MArrayDataHandle  mArrayDataHandle;
+    MArrayDataBuilder mBuilder;
+};
+} // namespace
+
+struct MayaUsdGeomNode::CacheData
+{
+    std::string                                loadedFile;
+    std::string                                rootPrim;
+    std::vector<UsdMaya_MeshDataReadJob::Data> primitives;
+};
+
+/* static */
+void* MayaUsdGeomNode::creator() { return new MayaUsdGeomNode(); }
+
+/* static */
+MStatus MayaUsdGeomNode::initialize()
+{
+    MStatus retValue = MS::kSuccess;
+
+    MFnTypedAttribute    typedAttrFn;
+    MFnCompoundAttribute compAttrFn;
+
+    // Inputs
+    filePathAttr
+        = typedAttrFn.create("filePath", "fp", MFnData::kString, MObject::kNullObj, &retValue);
+    typedAttrFn.setWritable(true);
+    typedAttrFn.setReadable(false);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    retValue = addAttribute(filePathAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    rootPrimAttr
+        = typedAttrFn.create("rootPrim", "rp", MFnData::kString, MObject::kNullObj, &retValue);
+    typedAttrFn.setWritable(true);
+    typedAttrFn.setReadable(false);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    retValue = addAttribute(rootPrimAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    // Outputs
+    outGeomAttr
+        = typedAttrFn.create("geometry", "geo", MFnData::kMesh, MObject::kNullObj, &retValue);
+    typedAttrFn.setWritable(false);
+    typedAttrFn.setReadable(true);
+    typedAttrFn.setStorable(true);
+    typedAttrFn.setArray(true);
+    typedAttrFn.setUsesArrayDataBuilder(true);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    retValue = addAttribute(outGeomAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    outGeomMatrixAttr
+        = typedAttrFn.create("matrix", "tra", MFnData::kMatrix, MObject::kNullObj, &retValue);
+    typedAttrFn.setWritable(false);
+    typedAttrFn.setReadable(true);
+    typedAttrFn.setStorable(true);
+    typedAttrFn.setArray(true);
+    typedAttrFn.setUsesArrayDataBuilder(true);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    retValue = addAttribute(outGeomMatrixAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    //
+    // add attribute dependencies
+    //
+    retValue = attributeAffects(filePathAttr, outGeomAttr);
+    retValue = attributeAffects(rootPrimAttr, outGeomAttr);
+    retValue = attributeAffects(filePathAttr, outGeomMatrixAttr);
+    retValue = attributeAffects(rootPrimAttr, outGeomMatrixAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    return retValue;
+}
+
+MayaUsdGeomNode::MayaUsdGeomNode()
+    : MPxNode()
+    , _cache(std::make_unique<CacheData>())
+{
+    TfRegistryManager::GetInstance().SubscribeTo<MayaUsdGeomNode>();
```

Comment:
I work on USD and I don't even know what this does. A comment welcome if you know.